### PR TITLE
Fix the User TimeZoneService

### PR DESCRIPTION
**Pull request by MikeAlhayek**
#16932

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs [hunk 1]
```diff
@@ -11,11 +11,27 @@ public interface IUserTimeZoneService
...
+    /// </summary>
+    /// <param name="userName">The user name.</param>
+    ValueTask<ITimeZone> GetAsync(string userName)
+        => ValueTask.FromResult<ITimeZone>(null);
+
+    /// <summary>
+    /// Forget the time zone for the given user.
+    /// </summary>
+    /// <param name="userName">The user name.</param>
+    ValueTask ForgetCacheAsync(string userName)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:35]: This assumes implementations will cache. It feels very weird, I don't think we have done this before.

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs [hunk 2]
```diff
@@ -11,11 +11,27 @@ public interface IUserTimeZoneService
...
     /// </summary>
     /// <param name="user">The <see cref="IUser"/>.</param>
+    [Obsolete("This method is obsolete and will be removed in future release. Instead use ForgetCacheAsync(userName).")]
     ValueTask UpdateAsync(IUser user);
+
+    /// <summary>
+    /// Gets the time zone for the given user.
+    /// </summary>
+    /// <param name="userName">The user name.</param>
+    ValueTask<ITimeZone> GetAsync(string userName)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:28]: This is not what we said we would, you even had some points to not do that. But move the caching at the timezone selector level.

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs (hunk 1 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:35]: We have other ways to manage cache invalidation. It feels we are going backwards from the initial issue we were fixing.

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs (hunk 2 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:28]: after looking more at the implemenation, I the UserTimeZoneService service becomes useless if we don't handle the cache in there. The UpdateAsync method suggest that it updates the timezone, but it does not do that ever, All it does is invalidate the cache. So it's best to pass the username to the GetAsync method and let it be a wrapper which handles the caching internally incause someone want to use it outside the timeZoneSelector.

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:35]: Currently the UpdateAsync(IUser user) invalidate the cache.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs:35]: @sebastienros I don't think we can tag a cache used by the distributed cache which will invalidate the cache based on the user tag. Do you have an idea in mind to auto invalidate the cache so we don't need such a method here?

**Comment by hishamco** [comment]
FYI I just wrote a comment [here](https://github.com/OrchardCMS/OrchardCore/pull/15796#issuecomment-2447909610) to allow @KeithRaven to chime again on his PR after there are a few comments later

I suggest closing this and letting Keith finish his PR unless he doesn't have time. I hope you understand what I mean @MikeAlhayek

File: src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Users/TimeZone/Services/IUserTimeZoneService.cs]: @sebastienros I actually removed the ForgetCacheAsync method from the interface. Now this interface needs a single method to get the time zone. I implemented the `IUserEventHandler` to invalidate the cache automatically.

**Comment by MikeAlhayek** [comment]
@hishamco, I am looking at that PR now. That PR does not fix the problem this PR fixes.

**Comment by hishamco** [comment]
I never mind creating another PR, at least we need to know the status of the previous PR, and why it's closed. Could the author responds to the feedback

**Comment by KeithRaven** [comment]
> I never mind creating another PR, at least we need to know the status of the previous PR, and why it's closed. Could the author responds to the feedback

Sorry I was going to sort out https://github.com/OrchardCMS/OrchardCore/pull/15796 however it was mentioned that another issue needed resolving first https://github.com/OrchardCMS/OrchardCore/pull/15796#issuecomment-2450804807.

This PR looks to be covering the same thing and more though, so makes sense to keep it in a single change set, so happy to leave the other closed.

**Comment by MikeAlhayek** [comment]
@deanmarcussen I believe this will fix your issue. However, we @sebastienros  and I have two different opinions on the implementation. We need your opinion to determine the final implementation before we merge. 

@sebastienros suggest that we remove both `IUserTimeZoneService` and `UserTimeZoneService`. Instead, place the same code in the UserTimeZoneSelector. So that the `UserTimeZoneSelector` will be responsible of caching the timezone for the current user and we have one less service to deal with.

Personally, I think it is okay to keep `IUserTimeZoneService` and `UserTimeZoneService` to allow others to reuse this service instead of having to couple that logic in the  `UserTimeZoneSelector`.

The question here, would you keep or remove both `IUserTimeZoneService` and `UserTimeZoneService`?

**Comment by deanmarcussen** [comment]
I don't have a massively strong opinion either way.

We used the service because it was there, and largely because we needed to be able to differentiate between user time, site time, and utc, particularly being able to provide those values to javascript.

Because we work in a very time distributed world, we need to present to users 2-3 date times to users, to save sanity. 
For example, we have sites in EU data centers which will run jobs at specific times, but they are often supported by users in the US, so being able to present what time a job will run at in both those timezones is a valid use case.

So realistically the timezone selector option doesn't work as ideally for us, as its results don't specify exactly what that timezone refers to. I can register it manually, and run it manually.

Mostly I've opened tickets about this service, to improve them for the community, and if the service goes away, then we will just adopt it internally, and manage it outselves.

I personally think it's a useful service, and concept (anyone implementing it themselves will probably forget to cache it, and drop their performance down, without knowing why)

**Comment by MikeAlhayek** [comment]
@sebastienros it sounds to me that @deanmarcussen is also leaning toward keeping the service. Unless you have a strong opinion on removing the service and move it into the time zone selector, I suggest we leave it.

**Comment by sebastienros** [comment]
Cool then @MikeAlhayek you can merge.

I wondered if we should keep `IUser` but in the end the service can always load it if it needs it.

**Pull request merged**